background/rust.tex

```diff
@@ -129,11 +130,15 @@ \subsubsection{Enums}
 \end{listing}
 
 Every \code{Option} variable is an {\enum} that can either have the algebraic value \code{Some} or the \emph{named value} \code{None}.
-If the variable is \code{Some}, the definition also says that it needs to contain another type \code{T}.
-This is also an example of generics in {\rust}, as \code{T} can be anything, e.g. an integer, a string, or a user defined type.
-The \code{Option} type is analogous with a normal {\C} pointer value, which either has the value 0, or \code{null} (analogous to \code{None}), or a value that points to a structure at a place in memory (analogous to \code{Some}).
-Even though the \code{Option} is analogous to a {\C} pointer, it is important to note their differences.
-It acts more like a wrapper around a potential value, where its internal value is not \emph{directly} accessible, which encourages the programmer to handle the cases where it might be \code{None}.
+If the variable is \code{Some}, the definition also says that it needs to contain a value of type \code{T}.
+This \code{T} can be anything and is how {\rust} defines a generic type.
+The \code{Option} type \emph{can} be used to encode the same programming pattern as \emph{pointers with a sentinel value of 0} (e.g. 0 for representing no value or the end of a datastructure), but in a fundamentaly safer way.
+The \emph{null-pointer dereference} problem caused by using this pattern is called the \emph{billion dollar bug} by its inventor Tony Hoare.
```

Comment:
Ja, jeg så etter det en gang jeg også. Men jeg husker fra Marius sin
oppgave at han hadde en del wiki-referanser

On Fri, Jun 12, 2015, 21:59 Håvard Wormdal Høiby <notifications@github.com>
wrote:

> In background/rust.tex
> <https://github.com/sondrele/thesis/pull/96#discussion_r32349469>:
>
> > @@ -129,11 +130,15 @@ \subsubsection{Enums}
> >  \end{listing}
> >
> >  Every \code{Option} variable is an {\enum} that can either have the algebraic value \code{Some} or the \emph{named value} \code{None}.
> > -If the variable is \code{Some}, the definition also says that it needs to contain another type \code{T}.
> > -This is also an example of generics in {\rust}, as \code{T} can be anything, e.g. an integer, a string, or a user defined type.
> > -The \code{Option} type is analogous with a normal {\C} pointer value, which either has the value 0, or \code{null} (analogous to \code{None}), or a value that points to a structure at a place in memory (analogous to \code{Some}).
> > -Even though the \code{Option} is analogous to a {\C} pointer, it is important to note their differences.
> > -It acts more like a wrapper around a potential value, where its internal value is not \emph{directly} accessible, which encourages the programmer to handle the cases where it might be \code{None}.
> > +If the variable is \code{Some}, the definition also says that it needs to contain a value of type \code{T}.
> > +This \code{T} can be anything and is how {\rust} defines a generic type.
> > +The \code{Option} type \emph{can} be used to encode the same programming pattern as \emph{pointers with a sentinel value of 0} (e.g. 0 for representing no value or the end of a datastructure), but in a fundamentaly safer way.
> > +The \emph{null-pointer dereference} problem caused by using this pattern is called the \emph{billion dollar bug} by its inventor Tony Hoare.
>
> Ah, drev jo å søkte, det finnes vel noen bedre artikler enn wiki.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sondrele/thesis/pull/96/files#r32349469>.
>